### PR TITLE
refactor(rust/tree_shaking): simplify detection on pure annotations

### PR DESCRIPTION
**Pull request by hyf0**
### Description

This PR also fixes:

- Support `/*@__PURE__*/` or `/*#__PURE__*/`.

### CODE REVIEW

File: crates/rolldown/tests/fixtures/tree_shaking/pure_annotation/artifacts.snap [hunk 1]
```diff
@@ -16,11 +16,6 @@ function test() {
 }
 const c = test();
 
-// d.js
```
Comment (hyf0) [inline crates/rolldown/tests/fixtures/tree_shaking/pure_annotation/artifacts.snap]: The result is the same as rollup for `d.js`.

https://rollupjs.org/repl/?version=4.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmQlN0QlMjBmcm9tJTIwJy4lMkZkLmpzJyU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwdGVzdDIoKSUyMCU3QiU3RCU1Q24lNUNuZXhwb3J0JTIwY29uc3QlMjBkJTIwJTNEJTIwJTJGKiUyMCUyM19fUFVSRV9fJTIwKiUyRiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHRlc3QyKCklMkMlMjBlJTIwJTNEJTIwJTIwdGVzdDIoKSUzQiU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIyZC5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 2]
```diff
@@ -36,22 +34,39 @@ static SIDE_EFFECT_FREE_MEMBER_EXPR_3: Lazy<FxHashSet<(&'static str, &'static st
...
+  ///
+  /// The only exception are parentheses that wrap a call or invocation.
+  ///
+  /// <https://rollupjs.org/configuration-options/#pure>
+  /// Copied from https://github.com/oxc-project/oxc/blob/147864cfeb112df526bb83d5b8671b465c005066/crates/oxc_linter/src/utils/tree_shaking.rs#L162-L171
+  pub fn has_pure_notation(&self, span: Span) -> bool {
+    let Some((start, comment)) = self.trivias.comments_range(..span.start).next_back() else {
+      return false;
+    };
+    let span = Span::new(*start, comment.end);
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: We could cache the return value of `contains(...)` per `comment.end`. This should do the optimization that these code did. But I wonder the if it's worth.

https://github.com/rolldown/rolldown/blob/9e6437d34e24c0c8f913009ebbb1b25fe4d9a3ce/crates/rolldown/src/ast_scanner/impl_visit.rs#L14-L27

---

If this optimization is worthy, we could create a good first issue for it or implement it in another PR.

File: crates/rolldown/src/ast_scanner/impl_visit.rs [hunk 3]
```diff
@@ -11,37 +11,10 @@ use super::{side_effect_detector::SideEffectDetector, AstScanner};
...
-    // the third element of tuple is whether the comment has been attached to any statement before,
-    // we use the third element and binary search, making the search fast e.g.
-    // ```js
-    // /** test**/
-    // call1();
-    // call2();
-    // ```
-    // Although, the `test` comment is both closest leading comment of `call1()` and `call2()`,
-    // but when we visited `call1()`, the second element of tuple will be set to true, so when we visited `call2()`,
-    // we don't need to check again;
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/impl_visit.rs:25]: Not if this PR needs this optimization too, need IWANABETHATGUY's confirmation.

**Comment by IWANABETHATGUY** [comment]
What do you mean by support `\*@__PURE__*\`, I don't see any usage in esbuild and rollup codebase,

**Comment by hyf0** [comment]
> What do you mean by support `\*@__PURE__*\`, I don't see any usage in esbuild and rollup codebase,

I made some typos. Should be `/* @__PURE__ */`

**Comment by IWANABETHATGUY** [comment]
> > What do you mean by support `\*@__PURE__*\`, I don't see any usage in esbuild and rollup codebase,
> 
> I made some typos. Should be `/* @__PURE__ */`

This two annotation should be supported in the previous pr, if I understand correctly, this pull request would support 
single comment `pure annotation`

**Review by IWANABETHATGUY — APPROVED** [review]

**Comment by hyf0** [comment]
> > > What do you mean by support `\*@__PURE__*\`, I don't see any usage in esbuild and rollup codebase,
> > 
> > 
> > I made some typos. Should be `/* @__PURE__ */`
> 
> This two annotation should be supported in the previous pr, if I understand correctly, this pull request would support single comment `pure annotation`

Yeah. Misunderstand the previous code.

**Pull request merged**